### ISSUE TITLE
ngc failed: Supplied parameters do not match any signature of call target.

### ISSUE THREAD

**Issue by johnking** [500tech/angular2-tree-component#97]
Hi there,

I have integrated the `angular2-tree-component` into my ionic-2 application, and my App works fine with browser testing `$ionic serve`.

but when I tried to load the app into my iOS device, I am experiencing an issue: `ngc building error` as below. 
Basically, it will cause `ngc` buidling error when using directive `<Tree ..></Tree>`. 

```
$ ionic build
...
[16:12:08]  Error: Error at /Users/johnking/workshop/ionic/wisdom/.tmp/node_modules/angular2-tree-component/dist/components/tree-node-content.component.ngfactory.ts:121:40
[16:12:08]  Supplied parameters do not match any signature of call target.
[16:12:08]  Error at /Users/johnking/workshop/ionic/wisdom/.tmp/node_modules/angular2-tree-component/dist/components/loading.component.ngfactory.ts:108:40
[16:12:08]  Supplied parameters do not match any signature of call target.
[16:12:08]  ngc failed
[16:12:08]  ionic-app-script task: "build"
[16:12:08]  Error: Error
...
```
the error indicates: `Supplied parameters do not match any signature of call target.`, and I can simply repeat the error by just using the simple example documented at: https://angular2-tree.readme.io/docs/getting-started. 

Below is the code for repeating the building error
```
import { Component} from '@angular/core';
@Component({
  selector: 'page-kepan',
  template: '<Tree [nodes]="nodes"></Tree>'
})
export class KepanPage {
  nodes = [
    {
      id: 1,
      name: 'root1',
      children: [
        { id: 2, name: 'child1' },
        { id: 3, name: 'child2' }
      ]
    },
    {
      id: 4,
      name: 'root2',
      children: [
        { id: 5, name: 'child2.1' },
        {
          id: 6,
          name: 'child2.2',
          children: [
            { id: 7, name: 'subsub' }
          ]
        }
      ]
    }
  ];
}

```

Based on the simple example above, it will cause `ngc` buidling error when using directive `<Tree ..></Tree>`. 

I looked into the files generated by the Angular 2 template compiler, both errors seem the same:

1. `tree-node-content.component.ngfactory.ts:121:40`
```
  detectChangesInternal(throwOnChange:boolean):void {
    var changes:{[key: string]:import7.SimpleChange} = (null as any);
    const currVal_0:boolean = !this.context.treeNodeContentTemplate;
    if (import4.checkBinding(throwOnChange,this._expr_0,currVal_0)) {
      this._NgIf_0_6.ngIf = currVal_0;
      this._expr_0 = currVal_0;
    }
    changes = (null as any);
    const currVal_1:any = this._map_0(this.context.node);
    if (import4.checkBinding(throwOnChange,this._expr_1,currVal_1)) {
      this._NgTemplateOutlet_2_6.ngOutletContext = currVal_1;
      if ((changes === (null as any))) { (changes = {}); }
      changes['ngOutletContext'] = new import7.SimpleChange(this._expr_1,currVal_1);
      this._expr_1 = currVal_1;
    }
    const currVal_2:any = this.context.treeNodeContentTemplate;
    if (import4.checkBinding(throwOnChange,this._expr_2,currVal_2)) {
      this._NgTemplateOutlet_2_6.ngTemplateOutlet = currVal_2;
      if ((changes === (null as any))) { (changes = {}); }
      changes['ngTemplateOutlet'] = new import7.SimpleChange(this._expr_2,currVal_2);
      this._expr_2 = currVal_2;
    }
    if ((changes !== (null as any))) { this._NgTemplateOutlet_2_6.ngOnChanges(changes); }
    this.detectContentChildrenChanges(throwOnChange);
    this.detectViewChildrenChanges(throwOnChange);
  }
}
```
  - line 121 is: `if ((changes !== (null as any))) `

2. `loading.component.ngfactory.ts:108:40`

```
  detectChangesInternal(throwOnChange:boolean):void {
    var changes:{[key: string]:import7.SimpleChange} = (null as any);
    const currVal_0:boolean = !this.context.loadingTemplate;
    if (import4.checkBinding(throwOnChange,this._expr_0,currVal_0)) {
      this._NgIf_0_6.ngIf = currVal_0;
      this._expr_0 = currVal_0;
    }
    changes = (null as any);
    const currVal_1:any = this.context.loadingTemplate;
    if (import4.checkBinding(throwOnChange,this._expr_1,currVal_1)) {
      this._NgTemplateOutlet_2_6.ngTemplateOutlet = currVal_1;
      if ((changes === (null as any))) { (changes = {}); }
      changes['ngTemplateOutlet'] = new import7.SimpleChange(this._expr_1,currVal_1);
      this._expr_1 = currVal_1;
    }
    if ((changes !== (null as any))) { this._NgTemplateOutlet_2_6.ngOnChanges(changes); }
    this.detectContentChildrenChanges(throwOnChange);
    this.detectViewChildrenChanges(throwOnChange);
  }
}
``` 
line 108 is: `if ((changes !== (null as any))) {` in the code above.

Thanks for any comments and suggestions to work around this issue. 

-John

**Comment by johnking** [comment]
@adamkleingit and @500tech-user 

I tried to find the help from ionic forum, but all the solutions are to make sure the function call to match with library API(s). But in my case, I only use `<Tree> </Tree> ` directive, so I have no idea which function call inside the directive `Tree` causing the building errors.  

May you please help on it? I wonder if we need to look into the `*.d.ts` files to make sure the API(s) exposed matching the library definitions? 

thanks 

-John

**Comment by johnking** [comment]
Spent days in trouble-shooting this issue.  

Finally, found the fact below. 

There was a bug in Angular2/ngTemplateOutlet, [Aot Error, Consider making ngOnChanges argument optional #11733](https://github.com/angular/angular/issues/11733)

I have to migrate the patch manually into my project since Ionic2 is not compatible with the latest version of Angular2.

**Comment by jskrzypek** [comment]
@johnking  I'm still getting this error with versions of angular & ionic that include that PR/commit. Are able to test again on a project where you aren't patching manually?

**Comment by jskrzypek** [comment]
Nvm, this was apparently fixed in angular2-tree-component@2.1.0

**Comment by GFoley83** [comment]
@johnking Did you manage to apply styles to the tree to make it look similar to the native Ionic components?
